### PR TITLE
8261211: Zero: alpha builds fail with '-fstack-protector not supported for this target'

### PR DESCRIPTION
**Pull request by shipilev**
```
Configuration summary:
* Name: linux-alpha-zero-slowdebug
* Debug level: slowdebug
* HS debug level: debug
* JVM variants: zero
* JVM features: zero: 'epsilongc g1gc jni-check jvmti management nmt parallelgc serialgc services vm-structs zero'
* OpenJDK target: OS: linux, CPU architecture: alpha, address length: 64
* Version string: 17-internal+0-adhoc.shade.jdk (17-internal)

Tools summary:
* Boot JDK: openjdk version "15-testing" 2020-09-15 OpenJDK Runtime Environment (build 15-testing+0-builds.shipilev.net-openjdk-jdk15-b17-20201010) OpenJDK 64-Bit Server VM (build 15-testing+0-builds.shipilev.net-openjdk-jdk15-b17-20201010, mixed mode, sharing) (at /home/shade/Install/jdk15.0.2-ea)
* Toolchain: gcc (GNU Compiler Collection)
* C Compiler: Version 9.3.0 (at /usr/bin/alpha-linux-gnu-gcc)
* C++ Compiler: Version 9.3.0 (at /usr/bin/alpha-linux-gnu-g++)

$ CONF=linux-alpha-zero-fastdebug make hotspot

* For target hotspot_variant-zero_libjvm_objs_cardTable.o:
cc1plus: error: '-fstack-protector' not supported for this target [-Werror]
cc1plus: all warnings being treated as errors
```

JDK-8199405 removed the conditional checks for -fstack-protector availability. I believe the intent was to trust that every recent GCC has this option. But this issue shows that stack-protector support is also arch-specific?

Linux x86_64 server configure now says:

```
checking if CC supports "-fstack-protector -Werror"... yes
checking if CXX supports "-fstack-protector -Werror"... yes
checking if both CC and CXX support "-fstack-protector -Werror"... yes
```

Linux Zero alpha configure now says:

```
checking if CC supports "-fstack-protector -Werror"... no
checking if CXX supports "-fstack-protector -Werror"... no
checking if both CC and CXX support "-fstack-protector -Werror"... no
```

Additional testing:
 - [x] Linux Zero alpha {release,fastdebug,slowdebug} cross-compilation
 - [x] Eyeballing configure logs for Linux Zero {alpha, x86_64}

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8261211](https://bugs.openjdk.java.net/browse/JDK-8261211): Zero: alpha builds fail with '-fstack-protector not supported for this target'

### Reviewers
 * [Magnus Ihse Bursie](https://openjdk.java.net/census#ihse) (@magicus - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2423/head:pull/2423`
`$ git checkout pull/2423`

### CODE REVIEW

**Comment by magicus** [comment]
> I believe the intent was to trust that every recent GCC has this option. But this issue shows that stack-protector support is also arch-specific?

Apparently so. But then again all platforms except alpha has this in recent gcc versions, so I think it was an understandable misunderstanding. :) Do anyone build alpha at any regularity at all? I thought Debian at least did that. But this must have been broken for quite some time?

File: make/autoconf/flags-cflags.m4 [hunk 1]
```diff
@@ -204,10 +204,15 @@ AC_DEFUN([FLAGS_SETUP_QUALITY_CHECKS],
       # This is most likely not really correct.
 
       # Add runtime stack smashing and undefined behavior checks.
-      CFLAGS_DEBUG_OPTIONS="-fstack-protector-all --param ssp-buffer-size=1"
-      CXXFLAGS_DEBUG_OPTIONS="-fstack-protector-all --param ssp-buffer-size=1"
+      # Not all versions of gcc support -fstack-protector-all
+      STACK_PROTECTOR_ALL_CFLAG="-fstack-protector-all"
+      FLAGS_COMPILER_CHECK_ARGUMENTS(ARGUMENT: [STACK_PROTECTOR_ALL_CFLAG -Werror],
```
Comment (magicus) [inline make/autoconf/flags-cflags.m4:209]: You don't need to add `-Werror` anymore, `$CFLAGS_WARNINGS_ARE_ERRORS` is now added automatically by `FLAGS_COMPILER_CHECK_ARGUMENTS` (as it should have been from the beginning).

File: make/autoconf/flags-cflags.m4 [hunk 2]
```diff
@@ -452,8 +457,16 @@ AC_DEFUN([FLAGS_SETUP_CFLAGS_HELPER],
...
 
   if test "x$TOOLCHAIN_TYPE" = xgcc; then
-    TOOLCHAIN_CFLAGS_JVM="$TOOLCHAIN_CFLAGS_JVM -fcheck-new -fstack-protector"
-    TOOLCHAIN_CFLAGS_JDK="-pipe -fstack-protector"
+    TOOLCHAIN_CFLAGS_JVM="$TOOLCHAIN_CFLAGS_JVM -fcheck-new"
+    TOOLCHAIN_CFLAGS_JDK="-pipe"
+
+    # Not all versions of gcc support -fstack-protector
+    STACK_PROTECTOR_CFLAG="-fstack-protector"
+    FLAGS_COMPILER_CHECK_ARGUMENTS(ARGUMENT: [$STACK_PROTECTOR_CFLAG -Werror],
```
Comment (magicus) [inline make/autoconf/flags-cflags.m4:465]: ... same here...

**Review by magicus — CHANGES_REQUESTED** [review]

File: make/autoconf/flags-cflags.m4 (hunk 1 above)
Comment (shipilev) [reply make/autoconf/flags-cflags.m4:209]: Fixed!

**Comment by shipilev** [comment]
> Apparently so. But then again all platforms except alpha has this in recent gcc versions, so I think it was an understandable misunderstanding. :) Do anyone build alpha at any regularity at all? I thought Debian at least did that. But this must have been broken for quite some time?

I discovered it after adding Alpha builds to my CIs, which are supposed to cover me for ongoing Zero maintenance work. So I would be building those regularly :) 

I just browsed Debian's build logs, and I believe Alpha builds are done with GCC 10 there (is that right, @glaubitz?). My CIs build with GCC 9. So maybe this patch would become unnecessary when we move to GCC 10 as default. I can work this around for Alpha builds in my CI, but I was hoping to have the least magical incantation to build cleanly.

**Review by magicus — APPROVED** [review]

**Comment by glaubitz** [comment]
> > Apparently so. But then again all platforms except alpha has this in recent gcc versions, so I think it was an understandable misunderstanding. :) Do anyone build alpha at any regularity at all? I thought Debian at least did that. But this must have been broken for quite some time?

Yes, we're building Alpha regularly in Debian. Keep in mind that we have additional patches in Debian which may not have been upstreamed yet (in particular on patch for m68k due to the different alignment).

> I discovered it after adding Alpha builds to my CIs, which are supposed to cover me for ongoing Zero maintenance work. So I would be building those regularly :)

Awesome, thanks for adding Alpha to your CI. Much appreciated.

> I just browsed Debian's build logs, and I believe Alpha builds are done with GCC 10 there (is that right, @glaubitz?).

That is correct!

> My CIs build with GCC 9. So maybe this patch would become unnecessary when we move to GCC 10 as default. I can work this around for Alpha builds in my CI, but I was hoping to have the least magical incantation to build cleanly.

I think you should just add a hack to your CI to work around this issue until OpenJDK has moved to gcc-10.

Thanks again for helping with the less common architectures. Much much appreciated and I owe you a beverage of your choice at the next FOSS conference ;-).

**Comment by shipilev** [comment]
> I think you should just add a hack to your CI to work around this issue until OpenJDK has moved to gcc-10.

Right. `--disable-warnings-as-errors` seems to be working as the workaround. I'll close this issue as WNF, instead of hacking the build system for this obscure corner case.

**Pull request closed without merging**